### PR TITLE
Expand pm.Data capacities

### LINKED ISSUE #3813: pymc3.Data converts input data to float64 type - so int data cannot later be used as an index
Hi there guys

I'd like to create a model that I want to fit many times to different datasets for cross validation purposes.

One of my columns of input data is categorical, so I use it to index a vector of RVs depending on which category is presented in each sample of data.  Something like this:

```
cat_mvs = [pymc3.Normal(c, mu = 0, sd = 0.05) for c in unique_categories]

cat_mv_vector = pymc3.math.stack(cat_mvs)
cat_data = pymc3.Data('Categorical Input Data', category_codes)
sample_cat_mv = cat_mv_vector[cat_data]
```

Note that my `category_codes` data is a numpy array of *integers*

That last line of code above triggers an error, here's the traceback within pymc3:

```
~/.pyenv/versions/3.7.2/envs/jupyterlab-3.7.2/lib/python3.7/site-packages/theano/tensor/var.py in __getitem__(self, args)
    568                             TensorVariable, TensorConstant,
    569                             theano.tensor.sharedvar.TensorSharedVariable))):
--> 570                 return self.take(args[axis], axis)
    571             else:
    572                 return theano.tensor.subtensor.advanced_subtensor(self, *args)

~/.pyenv/versions/3.7.2/envs/jupyterlab-3.7.2/lib/python3.7/site-packages/theano/tensor/var.py in take(self, indices, axis, mode)
    612 
    613     def take(self, indices, axis=None, mode='raise'):
--> 614         return theano.tensor.subtensor.take(self, indices, axis, mode)
    615 
    616     # COPYING

~/.pyenv/versions/3.7.2/envs/jupyterlab-3.7.2/lib/python3.7/site-packages/theano/tensor/subtensor.py in take(a, indices, axis, mode)
   2448             return advanced_subtensor1(a.flatten(), indices)
   2449         elif axis == 0:
-> 2450             return advanced_subtensor1(a, indices)
   2451         else:
   2452             if axis < 0:

~/.pyenv/versions/3.7.2/envs/jupyterlab-3.7.2/lib/python3.7/site-packages/theano/gof/op.py in __call__(self, *inputs, **kwargs)
    613         """
    614         return_list = kwargs.pop('return_list', False)
--> 615         node = self.make_node(*inputs, **kwargs)
    616 
    617         if config.compute_test_value != 'off':

~/.pyenv/versions/3.7.2/envs/jupyterlab-3.7.2/lib/python3.7/site-packages/theano/tensor/subtensor.py in make_node(self, x, ilist)
   1701         ilist_ = theano.tensor.as_tensor_variable(ilist)
   1702         if ilist_.type.dtype not in theano.tensor.integer_dtypes:
-> 1703             raise TypeError('index must be integers')
   1704         if ilist_.type.ndim != 1:
   1705             raise TypeError('index must be vector')

TypeError: index must be integers
```

It seems that within pymc3.Data(), my *category_codes* data is being coerced to float64, which is not a valid indexing type.

Looking at the source for `pymc3.Data()` I think the problem is ultimately in the called function `pymc3.model.pandas_to_array` which converts its input data to a float on its last line, see https://github.com/pymc-devs/pymc3/blob/master/pymc3/model.py#L1495

Can `pymc3.Data()` and/or `pymc3.model.pandas_to_array` be changed to be preserve the input data type?

Thanks!

### LINKED ISSUE #3842: Error using pm.Data to supply mu parameter to Normal
## Description of your problem

When I create a `pm.Data` with a vector of values for the mean of a Normal, I get a type error.  When I supply the same vector of values without packaging into `pm.Data`, this works fine.

**Please provide a minimal, self-contained, and reproducible example.**
```python
import numpy as np
import pymc3 as pm

# the following gives: TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
with pm.Model():
    x = pm.Data("x", [1.0, 2.0, 3.0])
    y = pm.Normal("y", mu=x, shape=3)

# same type error
with pm.Model():
    x = pm.Data("x", np.array([1.0, 2.0, 3.0]))
    y = pm.Normal("y", mu=x, shape=3)

# works fine
with pm.Model():
    x = np.array([1.0, 2.0, 3.0])
    y = pm.Normal("y", mu=x, shape=3)

```

**Please provide the full traceback.**
```python
>>> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/data_mu_bug.py", line 11, in <module>
  File "/Users/rpg/src/pymc3/pymc3/distributions/distribution.py", line 66, in __new__
    return model.Var(name, dist, data, total_size)
  File "/Users/rpg/src/pymc3/pymc3/model.py", line 998, in Var
    total_size=total_size, model=self)
  File "/Users/rpg/src/pymc3/pymc3/model.py", line 1450, in __init__
    distribution.shape, distribution.dtype) * distribution.default()
  File "/Users/rpg/src/pymc3/pymc3/distributions/distribution.py", line 91, in default
    return np.asarray(self.get_test_val(self.testval, self.defaults), self.dtype)
  File "/Users/rpg/src/pymc3/pymc3/distributions/distribution.py", line 96, in get_test_val
    if hasattr(self, v) and np.all(np.isfinite(self.getattr_value(v))):
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

**Please provide any additional information below.**

## Versions and main components

* PyMC3 Version: master
* Theano Version: 1.0.4
* Python Version: 3.6.8
* Operating system: MacOS
* How did you install PyMC3: pip, editable

### PR DESCRIPTION
**Pull request by AlexAndorra**
This PR expands on @hottwaj's work in #3816. Now, `pm.Data` can:
- Be used with integers as well as floats, which allows using it for index variables:
```python
with pm.Model() as model:
    index = pm.Data('index', np.array([2, 0, 1, 0, 2]))
    alpha = pm.Normal('alpha', 0, 1.5, shape=3)
    pm.Normal('obs', alpha[index], 1., observed=y)
```
- Be fed as input to other RVs, as asked in #3842:
```python
with pm.Model():
    x = pm.Data("x", [1.0, 2.0, 3.0])
    y = pm.Normal("y", mu=x, shape=3)
```
This is ready for review -- thanks a lot in advance for the reviews 🖖 

+ [x] No breaking changes (that I know of)
+ [x] Changes are covered by tests and docstrings
+ [x] Mentioned in the RELEASE-NOTES.md

### CODE REVIEW

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -14,6 +14,8 @@
 - `pm.sample` now has support for adapting dense mass matrix using `QuadPotentialFullAdapt` (see [#3596](https://github.com/pymc-devs/pymc3/pull/3596), [#3705](https://github.com/pymc-devs/pymc3/pull/3705), [#3858](https://github.com/pymc-devs/pymc3/pull/3858), and [#3893](https://github.com/pymc-devs/pymc3/pull/3893)). Use `init="adapt_full"` or `init="jitter+adapt_full"` to use.
 - `Moyal` distribution added (see [#3870](https://github.com/pymc-devs/pymc3/pull/3870)).
 - `pm.LKJCholeskyCov` now automatically computes and returns the unpacked Cholesky decomposition, the correlations and the standard deviations of the covariance matrix (see [#3881](https://github.com/pymc-devs/pymc3/pull/3881)).
+- `pm.Data` container can now be used for index variables, i.e with integer data and not only floats (see [#3925](https://github.com/pymc-devs/pymc3/pull/3925)).
```
Comment (lucianopaz) [inline RELEASE-NOTES.md]: Could you also link both of these to the corresponding issues?

File: pymc3/data.py [hunk 2]
```diff
@@ -478,27 +478,36 @@ class Data:
...
     """
-    def __new__(self, name, value):
+
+    def __new__(self, name, value, dtype=None):
+        if not dtype:
+            if hasattr(value, "dtype"):
+                # if no dtype given but available as attr of value, use that as dtype
+                dtype = value.dtype
+            elif isinstance(value, int):
+                dtype = int
```
Comment (lucianopaz) [inline pymc3/data.py]: It would be better to default to whatever `intX` would be instead.

File: pymc3/data.py [hunk 3]
```diff
@@ -478,27 +478,36 @@ class Data:
...
+            else:
+                # otherwise, assume float
+                dtype = float
```
Comment (lucianopaz) [inline pymc3/data.py]: It would be better to default to theano's `floatX`

**Review by lucianopaz — CHANGES_REQUESTED** [review]
Great work @alexandorra! I left some comments throughout the changed you made. What I think is the most important point that should be addressed is the default dtype. I think that it should be `theano.config.floatX`for floats and the corresponding `intX` dtype for ints. You can take a look at our `intX` function to find the mapping that we use to set the integer dtype based on theano's floatX.

**Review by michaelosthege — APPROVED** [review]

**Comment by michaelosthege** [comment]
I have nothing to add - Lucianos point on intX/float is important and the rest looks good 👍

**Comment by AlexAndorra** [comment]
Thanks for the reviews @lucianopaz and @michaelosthege ! 
Actually, while trying to implement Luciano's suggestions, I realized, quite surprisingly, that no type setting was needed to implement the "`pm.Data` as index variables" feature. Even more, explicitly setting the type with `floatX` and `intX` was failing everything 😮
You can see that in the updated code and tests -- please tell me if I'm mistaken because I'm really surprised by that result. Otherwise, I think this is ready to merge 😊

**Comment by michaelosthege** [comment]
> Thanks for the reviews @lucianopaz and @michaelosthege !
> Actually, while trying to implement Luciano's suggestions, I realized, quite surprisingly, that no type setting was needed to implement the "pm.Data as index variables" feature. Even more, explicitly setting the type with floatX and intX was failing everything 😮
> You can see that in the updated code and tests -- please tell me if I'm mistaken because I'm really surprised by that result. Otherwise, I think this is ready to merge 😊

You mean the pipeline for commit 0d07347 ?

It fails on a test that doesn't sound related:
`FAILED pymc3/tests/test_model_helpers.py::TestHelperFunc::test_pandas_to_array`

**Comment by AlexAndorra** [comment]
Ok, this was more complicated than I expected, but I think I understood the problem and figured out a more parcimonious implementation in the process. Tests pass locally -- if they pass on Travis too, this means it's ready for review 🍾

**Review by michaelosthege — APPROVED** [review]

**Review by lucianopaz — APPROVED** [review]
Great catch @AlexAndorra!

**Pull request merged**